2018-12-01/homework3.hs

```diff
@@ -0,0 +1,44 @@
+module Homework3 (parse) where
+
+import Data.Char (ord)
+
+data Pair a = Pair a a deriving Show
+
+parse :: String -> Either String (Pair Int)
+parse t = do
+    (t, r) <- parsePair (parseNumber 0) t
+    _ <- eod t
+    return r
+
+type Parser a = String -> Either String (String, a)
+
+consume :: (Char -> Either String a) -> Parser a
+consume _ "" = Left "unexpected end"
+consume f t = do
+    case f (head t) of
+        Right a -> Right ((tail t), a)
+        Left m -> Left m
+
+expect :: Char -> Parser ()
+expect c = consume (\x -> if x == c then Right () else Left ("expected '" ++ [c] ++ "', but got '" ++ [x] ++ "'"))
+
+eod :: Parser ()
+eod "" = Right ("", ())
+eod _ =  Left "tailing characters"
+
+parsePair :: Parser a -> Parser (Pair a)
+parsePair f t = do
+    (t, _) <- expect '(' t
+    (t, a) <- f t
+    (t, _) <- expect ',' t
+    (t, b) <- f t
+    (t, _) <- expect ')' t
+    return (t, (Pair a b))
+
+parseNumber :: Int -> Parser Int
```

Comment:
It's better to create a new combinator for extracting sequences of digits. Usually your "consume" parser would be called "satisfy". Then you can create a new combinator `many :: Parser a -> Parser [a]` that would apply the given parser until it can be applied, and then `parseNumber` will be something like `read ... many (satisfy isDigit)`.